Add rust compile time blog (#282)
pingcap/blog: rust-compile-time-adventures-with-tikv-ep1.md

---
title: "The Rust Compilation Model Calamity"
author: ['Brian Anderson']
date: 2020-01-29

> summary: In this first episode of the Rust Compile time series, Brian Anderson, co-author of Rust, shares with you his researches and experiences with Rust compile times, using the TiKV project as a case study.

Author: ```suggestion
summary: In this first episode of the Rust Compile time series, Brian Anderson, one of Rust's original authors, shares with you his researches and experiences with Rust compile times, using the TiKV project as a case study.
```

---

> title: Rust Compile-time Adventures with TiKV: Episode 1

Reviewer: ```suggestion
title: "Rust Compile-time Adventures with TiKV: Episode 1"
```

Reviewer: Because the title contains a colon, we need to add quotes for it in the metadata.

author: ['Brian Anderson']
date: 2020-01-29
summary: In this first episode of the Rust Compile time series, Brian Anderson, co-author of Rust, shares with you his researches and experiences with Rust compile times, using the TiKV project as a case study.
tags: ['Engineering']
categories: ['Engineering']
image: /images/blog/rust-compile-time-adventures.png
---

> _The Rust programming language was designed for slow compilation times._
>
> _I was there. I witnessed it for myself, and I am finally ready to break the silence: Rust is a hoax. It's a prank the language designers played on you, the Rust user — adopt this high-performance, high-reliability language for your products, and we'll reduce your productivity to a crawl._

> ![Rust Compile Time Adventures with TiKV](media/rust-compile-time-adventures.png)

Reviewer: I'm afraid the banner is a little small.

Other nodes in TiDB are written in Go, which of course comes with a different set of advantages and disadvantages from Rust. Some of the Go developers at PingCAP resent having to wait for the Rust components to build. They are used to a rapid build-test cycle.

Rust developers, on the other hand, are used to taking a lot of coffee breaks (or tea, cigarettes, sobbing, or whatever as the case may be — Rust developers have the spare time to nurse their demons).

### Preview: The TiKV Compile-time adventure so far

The first entry in this series is just a story about the history of Rust with respect to compilation time. Since it might take several more entries before we dive into concrete technical details of what we've done with TiKV's compile times, here's a pretty graph to capture your imagination, without comment.

![Rust Compile Time for TiKV](media/rust-compile-times-tikv.svg)

> 

Reviewer: <div class="caption-center"> Rust Compile Time for TiKV </div>

Reviewer: We need to add an image caption for each image to improve our SEO performance.

Here are some cursory examples of the Servo-Rust feedback loop:

* Labeled break and continue [was implemented in order to auto-generate an HTML parser](https://github.com/rust-lang/rust/issues/2216).
* Owned closures [were implemented after analyzing closure usage in Servo](https://github.com/rust-lang/rust/issues/2549#issuecomment-19588158).
* External function calls used to be considered safe. [This changed in part due to experience in Servo](https://github.com/rust-lang/rust/issues/2628#issuecomment-9384243).
* The migration from green-threading to native threading was informed by the experience of building Servo, observing the FFI overhead of Servo's SpiderMonkey integration, and profiling "hot splits", where the green thread stacks needed to be expanded and contracted.

The co-development of Rust and Servo created a [virtuous cycle](https://en.wikipedia.org/wiki/Virtuous_circle_and_vicious_circle) that allowed both projects to thrive. Today, Servo components are deeply integrated into Firefox, ensuring that Rust cannot die while Firefox lives.

Mission accomplished.

> ![Rust mission accomplished](media/rust-compile-mission-completed.png)

Reviewer: Is this image from the Internet source? If yes, we need to add the source here to avoid copyright issues.

Reviewer 2: The image comes from NASA so I believe it is in the public domain.

I downloaded it here: https://www.bbc.co.uk/radioscotland/60s/moonlandings/05

Author: Nice photoshop there!

* _Monomorphization_ — Rust translates each generic instantiation into its own machine code, creating code bloat and increasing compile time.

* _Stack unwinding_ — stack unwinding after unrecoverable exceptions traverses the callstack backwards and runs cleanup code. It requires lots of compile-time book-keeping and code generation.

* _Build scripts_ — build scripts allow arbitrary code to be run at compile-time, and pull in their own dependencies that need to be compiled. Their unknown side-effects and unknown inputs and outputs limit assumptions tools can make about them, which e.g. limits caching opportunities.

* _Macros_ — macros require multiple passes to expand, expand to often surprising amounts of hidden code, and impose limitations on partial parsing. Procedural macros have negative impacts similar to build scripts.

* _LLVM backend_ — LLVM produces good machine code, but runs relatively slowly.

* _Relying too much on the LLVM optimizer_ — Rust is well-known for generating a large quantity of LLVM IR and letting LLVM optimize it away. This is exacerbated by duplication from monomorphization.

> * _Split compiler / package manager_ — although it is normal for languages to have a package manager separate from the compiler, in Rust at least this results in both cargo and rustc having imperfect and redundant information about the overall compilation pipeline. As more parts of the pipeline are short-circuited for efficiency, more metadata needs to be transferred between instances of the compiler, mostly through the filesystem, which has overhead.

Reviewer: ```suggestion
* _Split compiler/package manager_ — although it is normal for languages to have a package manager separate from the compiler, in Rust at least this results in both cargo and rustc having imperfect and redundant information about the overall compilation pipeline. As more parts of the pipeline are short-circuited for efficiency, more metadata needs to be transferred between instances of the compiler, mostly through the filesystem, which has overhead.
```

![Rust Compile Time Adventures with TiKV](media/rust-compile-time-adventures.png)

## Rust compile times are bad

I know it because I encounter it every day. I know it because my coworkers complain about it. I know it because the Rust production users I talk to consistently mention it. I know it because so many of us indicated so [in the last Rust survey](https://blog.rust-lang.org/2018/11/27/Rust-survey-2018.html#challenges).

This is kinda infuriating, as almost everything that matters about Rust is pretty damn good. But Rust compile times are so, so bad.

Rust compile times are perhaps its biggest weakness.

At [PingCAP](https://pingcap.com/), we develop our distributed storage system, [TiKV](https://github.com/tikv/tikv/), in Rust, and it compiles slow enough to discourage many in the company from using Rust. I recently spent some time, along with several others on the TiKV team and its wider community, investigating TiKV's compile times.

Over a series of posts, I'll discuss what we have learned:

- Why compiling Rust is slow, and/or feels slow;

> - How Rust's development lead to slow compile-times;

Reviewer: ```suggestion
- How Rust's development leads to slow compile-times;
```

> - How Rust's development lead to slow compile-times;

Reviewer 2: This is correct as written. Past tense.
